test/jaxp/javax/xml/jaxp/unittest/xpath/XPathExpFollowingTest.java

```diff
@@ -0,0 +1,154 @@
+                {"/Customers/following-sibling::*", "/None"},
+                {"/Customers/Customer/following-sibling::Customer", "/Customers/Customer[@id != 'x1']"},
+                {"/Customers/Customer/following-sibling::foo:Customer", "/Customers/foo:Customer"},
+                {"/Customers/Customer[@id='x1']/Name/following-sibling::Address", "/Customers/Customer[@id='x1']/Address"},
+                {"/Customers/Customer/Name/following-sibling::Address", "/Customers//Address"},
+                {"/Customers/Customer[@id='x1']/following-sibling::*", "/Customers/Customer[@id != 'x1'] | /Customers/foo:Customer"},
+                {"/Customers/foo:Customer/foo:Address/following-sibling::*", "/Customers/foo:Customer/foo:Age | /Customers/foo:Customer/foo:ClubMember"}
+        };
+    }
+
+    /*
+     * DataProvider: provides XPath following expressions and expected number of following nodes returned
+     */
+    @DataProvider(name = "followingXpathNodeCount")
+    public Object[][] getFollowingXpathExpressionNodeCount() {
+        return new Object[][] {
+                {"/Customers/following::*", 0},
+                {"/Customers/Customer/following::*", 30},
+                {"/Customers/Customer/following::Customer", 2},
+                {"/Customers/Customer/following::foo:Customer", 1},
+                {"/Customers/Customer[@id='x1']/Name/following::*", 38},
+                {"/Customers/Customer/Address/following::*", 32},
+                {"/Customers/foo:Customer/foo:Address/following::*", 2},
+                {"/Customers/foo:Customer/foo:Name/following::*", 8},
+                {"/Customers/following-sibling::*", 0},
+                {"/Customers/Customer/following-sibling::*", 3},
+                {"/Customers/Customer/following-sibling::Customer", 2},
+                {"/Customers/Customer/following-sibling::foo:Customer", 1},
+                {"/Customers/Customer[@id='x1']/Name/following-sibling::*", 5},
+                {"/Customers/Customer/Address/following-sibling::*", 6},
+                {"/Customers/Customer[@id='x1']/Address/following-sibling::*", 2},
+                {"/Customers/foo:Customer/foo:Address/following-sibling::*", 2}
+        };
+    }
+
+    /*
+     * DataProvider: provides XPath following expressions which should not return any node.
+     */
+    @DataProvider(name = "followingXpathEmpty")
+    public Object[][] getFollowingXpathExpressionEmpty() {
+        return new Object[][] {
+                {"/Customers/following::*"},
+                {"/Customers/foo:Customer/following::*"},
+                {"/Customers/Customer[@id = 'x3' ]/following::Customer"},
+                {"/Customers/following::id"},
+                {"/Customers/Customer[@id = 'x3' ]/following-sibling::Customer"},
+                {"/Customers/foo:Customer/following-sibling::*"},
+                {"/Customers/Customer/following-sibling::foo:Name"},
+                {"/Customers/following-sibling::id"}
+        };
+    }
+
```

Comment:
It would be good to add a predicate test for following. e.g.
`//foo:Customer/*[not(following::*)]`
`//*[not(following::*) and not(preceding::*)]`
`//Street[following::Street and preceding::Street]`

Reply:
Also good for following-sibling.